### PR TITLE
Add Thread Duration overloads

### PR DESCRIPTION
**Pull request by He-Pin**
## Motivation
PR #4851 adds `Thread.sleep(Duration)` and `Thread.join(Duration)` support, which should be reviewed as a small follow-up to the minimal Duration API.

## Modification
- Use `Duration.toMillisAndNanos` for Thread sleep/join duration overloads.
- Preserve the required non-positive duration no-op behavior.
- Add JDK19-gated tests for the overload behavior.

## Result
Thread Duration overload behavior is available once the Duration foundation from #4863 is merged.

## Merge Order
- Stack position: 3, after #4863.
- Depends on: #4863, which depends on #4856.
- Parallel with: #4866; both depend on #4863 but do not depend on each other.
- Review note: until #4856 and #4863 merge, GitHub may show their parent diffs.
- After merge: rebase #4851 and drop this Thread duration diff.

## Verification
- `scripts/scalafmt --test`
- `sbt "javalib2_13/compile" "tests2_13/Test/compile"`

## References
- Split from #4851.
- Depends on #4863 and #4856.

### CODE REVIEW

**Comment by LeeTibbert** [comment]
Triage:  the`Thread` overloads deserve to be looked  at on a big screen. They may well be a good correction.

Other are probably not good candidates. Unit test location for JDK 19 features is not consistent with current SN practice. Also Duration test would need to be in tests-ext (??).


 Contents appeared redundant & in excess of need. I stopped reading before seeing a possibly useful Duration case. If such is there, what is its Intellectual Property status ( i.e. copied from OpenJDK?).

ChronoUnit deserves its own javalib-ext-dummies discussion and possible PR.

**Comment by He-Pin** [comment]
Updated this draft to follow the #4876 direction:

- removed the javalib `java.time.*` implementation from the PR diff
- kept the PR focused on `Thread.sleep(Duration)` / `Thread.join(Duration)`
- implemented conversion through `TimeUnit.NANOSECONDS.convert(Duration)`, matching the JDK path and reusing the ext-dummies support from #4876
- moved coverage to native-only `testsExt` so it can link against `javalib-ext-dummies` without requiring normal unit-tests to provide java.time classes

Local verification:
- `./scripts/scalafmt --test javalib/src/main/scala/java/lang/Thread.scala unit-tests-ext/native/src/test/scala/org/scalanative/testsuite/javalib/lang/ThreadDurationTest.scala javalib-ext-dummies/src/main/scala/java/time/DateTimeException.scala`
- `git diff --check`
- `sbt "javalib2_13/compile" "testsExt2_13/testOnly org.scalanative.testsuite.javalib.lang.ThreadDurationTest"`

The PR diff is now down to `Thread.scala` plus one native tests-ext file.

**Comment by He-Pin** [comment]
Moved the Duration overload tests out of tests-ext and into the existing JDK19-gated shared test source tree. The previous tests-ext placement made CI's JDK17 host compiler reject direct calls to Thread.sleep(Duration) / Thread.join(Duration) before Scala Native linking.

Local verification on JDK 25:
- ./scripts/scalafmt --test javalib/src/main/scala/java/lang/Thread.scala unit-tests/shared/src/test/require-jdk19/org/scalanative/testsuite/javalib/lang/ThreadDurationTestOnJDK19.scala
- git diff --check
- sbt "testsExt2_13/Test/compile"
- sbt "testsJVM2_13/testOnly org.scalanative.testsuite.javalib.lang.ThreadDurationTestOnJDK19"
- sbt "tests2_13/testOnly org.scalanative.testsuite.javalib.lang.ThreadDurationTestOnJDK19"

**Comment by LeeTibbert** [comment]
This PR has promise.

I'll take a closer look after the current Virtual Threads
work settles down.  Keeping out of the kitchen while the chief chef is working.

**Comment by LeeTibbert** [comment]
The three failing compliance checks are due to a link unknown symbol failing. Expected since Duration is used and   by design and intent unit-tests project does not supply a local or third party java.time implementation.

I believe the test needs to  be back in unit-tests-ext.
I think the does-not-compile on all JDKs problem here is the same or similar problem that lead me to the 
hack of requiring that testsExt be no-CI and manual,with all the difficulty that   brings. See top of that file.

testsExt does  not have  fully elaborated requires-JDK
environment. I'll study that issue when I study this PR.

Perhaps enough incentive is developing to create a
single testsExt require-JDK-LTS directory tree ONLY
and move timeUnitExt and ThreadExt there.
Another, but lesser, hack. Candidate for discussion in a other PR, after a private design study. 

One hack is unfortunate expediency, two raise concern, three require preventative action. Here I am sheepishly suggesting hack  number 2 and dreading the a other appearance of the situation leading to these hacks.
Further JDK 17/27 work may  bring us there,
say security classes or methods.


For now, I suggest  trial run of following the TimeUnitExt.scala.DISABLED after having run the test manually on JDK 8 or 17.

Others may have a different suggestion/requirement
so may be worth waiting a few days for concensus.

**Comment by LeeTibbert** [comment]
On a cleanup run I checked `Thread.scala.

Looks like the two methods of thos PR have been Implemented for three years. Am I missing something?

I take a look if this PR has tests and what unit-tests-ext might have.

**Comment by LeeTibbert** [comment]
Not sure why the LLM wants to change existing code.

One suggested change simplified two methods calls down to one <= 0. Probably worthwhile.

Other places I spit checked each lead a  cranky me to ask "Why?".

**Comment by LeeTibbert** [comment]
I think the state of this PR is that there is a unit test which should move to the appropriate sub-directory of
unit-tests-ext.

May have to add iszero and innovative methods to the Duration mock. Maybe less work to accept the <0 change of this PR.

SN world has changed since this PR was created, so OnJDK suffix can be dropped.

File: javalib/src/main/scala/java/time/DateTimeException.scala [hunk 1]
```diff
@@ -1,4 +1,3 @@
-// Ported from Scala.js, commit: 54648372, dated: 2020-09-24
```
Comment (coreyoconnor) [inline javalib/src/main/scala/java/time/DateTimeException.scala:1]: Is removing this attribution line standard?

File: javalib/src/main/scala/java/lang/Thread.scala [hunk 2]
```diff
@@ -378,6 +378,7 @@ class Thread private[lang] (
   )
```
Comment (coreyoconnor) [inline javalib/src/main/scala/java/lang/Thread.scala:378]: These changes can be their own PR: Fix of Thread handling of invalid duration arguments.

File: javalib/src/main/scala/java/util/concurrent/TimeUnit.scala [hunk 3]
```diff
@@ -160,7 +160,7 @@ object TimeUnit {
 
   private def x(a: Long, b: Long, max: Long): Long = {
     if (a > max) MAX
-    else if (a < -max) -MAX
+    else if (a < -max) MIN
```
Comment (coreyoconnor) [inline javalib/src/main/scala/java/util/concurrent/TimeUnit.scala:163]: This change can be it's own PR.

**Review by coreyoconnor — COMMENTED** [review]
Good changes overall. Several PRs here!

Unfortunate fact, IMO, the cost of splitting one PR into multiple is annoyingly high - even with LLMs. Shouldn't even require an LLM to support a request like: "split this components into their own PRs". Yet that is the best way to avoid the review bottleneck: Tiny PRs for the humans

File: javalib/src/main/scala/java/time/DateTimeException.scala (hunk 1 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/time/DateTimeException.scala:1]: @coreyoconnor 

No it is not, good catch.

For future  readers:

Usually the historical  recognition is, deservedly, left in place and some  brief comment a out Scala Native code additions or  modifications


Prompted by this question  I reviewed the entries of this PR. I think I was left kindly asking "Why make any change? What is the benefit?"

There might be some benefit to adding  the proposed unit-test in the proper "-ext" place.

Will look again more closely if anybody thinks I have missed something subtle that the LLM is wanting to do.

File: javalib/src/main/scala/java/util/concurrent/TimeUnit.scala (hunk 3 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/util/concurrent/TimeUnit.scala:163]: From looking at the SN code again just now
I suggest any one proposing to make this change write some tests of edge cases to show it is gives same results as current code.

Looks  like current code is trying to clamp(-MAX, a, MAX). Proposed change would change at low bound.

That may be the point of the proposed change. One would need to carefully check the use sites. 

I suspect the current idiom is equivalent to an abs() and the bit manipulations depend on no change in absolute magnitude.

I'm stuck in 2 or 3 other thickets right now
and am loathe to take an interrupt for a doubtful change.

Comment (coreyoconnor) [reply javalib/src/main/scala/java/util/concurrent/TimeUnit.scala:163]: Agreed. Sure, on the surface `MIN` seems appropriate. However I think you are right: The code is a fancy `clamp(-MAX, a, MAX)`. In which case the existing `-MAX` is expected.

Comment (LeeTibbert) [reply javalib/src/main/scala/java/util/concurrent/TimeUnit.scala:163]: probably from the days before SN, or even Java itself,  implemented `jl.Math.clamp().

**Comment by LeeTibbert** [comment]
@He-Pin  

I think this puppy can be closed, unless you can identify some benefit.  

Thanks,

Lee
